text/3727-unsafe-set-discriminant.md

```diff
@@ -0,0 +1,318 @@
+// copies by value, and the allocation is not local.
+let out: Option<[u32; 1024]> = unsafe { out.assume_init() };
+assert_eq!(out.is_some(), init_some.is_some());
+if let Some(val) = init_some {
+    assert!(out.as_ref().unwrap().iter().all(|x| *x == val));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `discriminant_of!` macro
+
+This macro is the "valueless" version of `std::mem::discriminant`. It needs to be a macro, as we cannot otherwise name the variant.
+
+Unlike the `discriminant` function, it is possible at compile time to detect if this is called on a type that is not an `enum`, meaning that it is possible for it to be a compiler error in this case, rather than "unspecified but not UB" to use this with a non-enum type.
+
+This macro returns by value the same opaque `Discriminant<T>` used by `std::mem::discriminant`.
+
+This RFC does not propose allowing obtaining the discriminant of "nested" fields, and will only work with a "top level" item, but does not preclude adding this capability in the future. This addition could be specified in a later RFC. For now users are recommended to handle "one level at a time".
+
+```rust
+discriminant_of!(Option<Option<u32>>, Some);        // Allowed
+discriminant_of!(Option<Option<u32>>, Some.0.Some); // Not Allowed
+```
+
+## The `set_discriminant` function
+
+This function must be unsafe, as setting an incorrect discriminant could lead to undefined behavior by allowing safe code to observe incorrectly or incompletely initialized values.
+
+This function takes an `*mut T` and has multiple requirements necessary for safety:
+
+* The pointer `*mut T` must be non-null
+* The enum `T` and the selected variant's fields must be well-aligned for reading and writing
+* This function is allowed to write and read-back both the discriminant and body (whether they each exist or not, and whether the discriminant and body are separate or not). This function also may do this in an unsynchronized manner (not requiring locks or atomic operations), which means an exclusive access is required
+
+If the `T` used for `*mut T` or `Discriminant<T>` when calling this function is NOT an enum, then this function is specified as a no-op. This is not undefined behavior, but later calls to `assume_init` may be undefined behavior if the `T` was not properly initialized. This is also allowed (but not required) to cause a debug assertion failure to aid in testing (implementor's choice).
+
+When this function is called, it MUST be called AFTER fully initializing the variant of the enum completely, as in some cases it may be necessary to read back these values. This is discussed more in the next section.
+
+Semantically, `set_discriminant` is specified to optionally write the discriminant (when necessary), and read-back the discriminant. If the read-back discriminant does not match the expected value, then the behavior is undefined.
```

Comment:
It'd be good also for this (or the following) section to be extended with more details about why this is the correct semantic. I think I see the case for and agree that setting a `MaybeUninit<Option<&u8>>` to `Some` with `set_discriminant` should represent an assertion that the relevant bits aren't zero, but it's natural to wonder about this, and so it'd be good to describe it in some detail.

Reply:
This sounds about right, though I would use some of the terms a bit differently. Specifically, set_discriminant will never read or write anything but the discriminant. However, enum layout is unspecified, and the discriminant can overlap with arbitrary values of the body.

It may be worth spelling out exactly what the safety requirements are: when writing the discriminant of variant `V`, then *all fields of `V` that overlap with the discriminant must be set to values that satisfy the language invariant of their type*. In practice, we don't guarantee which fields overlap with the discriminant, so *all* fields must be properly initialized.

This can replace the discussion of "read back" etc, that is just an implementation detail. Likewise, you can keep the example setting `None`, but I would frame it differently: because `None` has no fields, the requirement is trivially met. So the instruction is not "responsible for also initializing the whole body of the variant", all it really has to ensure is that future reads of the discriminant behave correctly.

> It'd be good also for this (or the following) section to be extended with more details about why this is the correct semantic.

The reason we introduced the readback is to make it so that
```
SetDiscriminant(place, variant_idx);
_x = GetDiscriminant(place);
```
can be optimize to
```
SetDiscriminant(place, variant_idx);
_x = variant_idx;
```
This seems like an obviously desirable transformation, so we must ensure that e.g. writing the `Some` discriminant can only return successfully if the next `GetDiscriminant` indeed would return `Some`.

Reply:
@RalfJung Would the optimization you have in mind be impeded if we merely say that any enum value which is not wrapped in `MaybeUninit` is valid? Ie, that initialization of the variant and fields may happen in any order, but that initialization must be complete by the time that `MaybeUninit::assume_init` is called? That would permit `set_discriminant` to not need to constrain the existing content of the enum since it would basically just be a fancy `ptr::write`, and it would make it easier to reason about for callers. I assume there's a reason that this simpler approach isn't viable, but I'm curious what that reason is.

Reply:
I don't understand the question. It seems you are proposing to just not do the optimization. But we think we should reserve the possibility to do that optimization, and then it follows immediately that we need the UB.

IMO it'd also be rather strange to say that setting the discriminant of `Option<&T>` to `Some` on an all-null value is just a NOP and permitted. The programmer asked us to do something nonsensical.